mne/time_frequency/psd.py

```diff
@@ -128,16 +143,29 @@ def psd_array_welch(x, sfreq, fmin=0, fmax=np.inf, n_fft=256, n_overlap=0,
                              for d in x_splits)
 
     # Combining, reducing windows and reshaping to original data shape
-    psds = np.concatenate([np.nanmean(f_s, axis=-1)
-                           for f_s in f_spectrogram], axis=0)
-    psds.shape = np.hstack([dshape, -1])
+    if combine is not None:
+        combine = _get_reduction(combine)
+        psds = np.concatenate([combine(f_s) for f_s in f_spectrogram],
+                              axis=0).reshape(dshape + (-1,))
+    else:
+        n_windows = psds.shape[-1]
+        psds = psds.reshape(dshape + (-1, n_windows))
+        # windows are put at -3 position so that the output is of shape
+        # windows x channels x frequencies for raw data and
+        # epochs x windows x channels x frequencies for epochs
+        if psds.ndim > 2:
+            # we assume that 2d input data are channels x times, if they are
+            # epochs x times the output is windows x epochs x freqs
```

Comment:
Basically just move this swap/rolling logic to the `Epochs` or `Raw` as necessary, and leave this function simple.